pkg/compactor/bucket_compactor.go

```diff
@@ -994,19 +997,31 @@ func (c *BucketCompactor) filterOwnJobs(jobs []*Job) ([]*Job, error) {
 	return jobs, nil
 }
 
-// filterJobsByWaitPeriod filters out jobs for which the configured wait period hasn't been honored yet.
-func (c *BucketCompactor) filterJobsByWaitPeriod(ctx context.Context, jobs []*Job) []*Job {
+// filterFirstLevelJobs filters out jobs for which the configured compaction delay or wait period have not yet elapsed.
+// When a compactionDelay is configured, this setting will be used, else the waitPeriod will be used.
+func (c *BucketCompactor) filterFirstLevelJobs(ctx context.Context, jobs []*Job) []*Job {
 	for i := 0; i < len(jobs); {
-		if elapsed, notElapsedBlock, err := jobWaitPeriodElapsed(ctx, jobs[i], c.waitPeriod, c.bkt); err != nil {
-			level.Warn(c.logger).Log("msg", "not enforcing compaction wait period because the check if compaction job contains recently uploaded blocks has failed", "groupKey", jobs[i].Key(), "err", err)
-
-			// Keep the job.
-			i++
-		} else if !elapsed {
-			level.Info(c.logger).Log("msg", "skipping compaction job because blocks in this job were uploaded too recently (within wait period)", "groupKey", jobs[i].Key(), "waitPeriodNotElapsedFor", notElapsedBlock.String())
-			jobs = append(jobs[:i], jobs[i+1:]...)
+		// Filter jobs using compaction delay when enabled
+		if c.compactionDelay > 0 {
```

Comment:
I read the issue https://github.com/grafana/mimir/issues/7279 and I think we can simply exclude OOO blocks from the check done by `jobWaitPeriodElapsed()`. In `jobWaitPeriodElapsed()` you can skip the check for blocks having `meta.OutOfOrder == true`.

WDYT?

Reply:
That would help, but blocks can be uploaded late for other reasons: one described in #7306, or just when ingesters are busy. Perhaps excluding OOO and fixing #7306 would resolve most of them though.

Reply:
I'm wondering if we should treat #7306 differently: yesterday we talked on Slack about changing `Head.compactable()`. Once changed as discussed, would #7306 still be a concern for #7279?

Reply:
Fixing #7306 along with the OOO check you suggested above should solve the problems I hit with #7279. It's possible blocks might be uploaded late for reasons that #7306 doesn't address, like overloaded ingesters, in which case we still might need an approach like this PR takes, but I didn't see overloaded ingesters causing late blocks in recent testing.

So it probably makes sense to go with the simple OOO check for #7306 and see how that goes, along with fixing #7279 as discussed. I can close this PR and followup with those separately.

Reply:
Opened:

https://github.com/grafana/mimir-prometheus/pull/597
https://github.com/grafana/mimir/pull/7342